3.next - Add Cookie encryption middleware

markstory opened the pull request:
In order to deprecate the CookieComponent, we need to have an alternative solution. By having middleware that transparently encrypts/decrypts cookies we can deprecate the CookieComponent.

The previous use cases of using different encryption keys for different cookie names, can be afforded by applying the EncryptedCookieMiddleware multiple times.

### TODO

* [x] Complete API docs

### Questions

* Does the whitelist approach make sense?
* Should there be a `*` mode to encrypt all cookies?
* Are there other features of CookieComponent that are missing between the new Cookie objects and this middleware?

Refs #10406

markstory committed:
Sketch out EncryptedCookieMiddleware

Include support for Set-Cookie and cookie objects being encoded. Still
need lots of docs and tests.

markstory committed:
Cookies should be % encoded, not + encoded.

markstory committed:
Add outline of tests for EncryptedCookieMiddleware.

markstory committed:
Fix phpcs.

burzum commented:
The whitelist approach is OK. The * is also OK, I think it's a very rare case somebody encrypts multiple cookies differently? 

Something missing... not sure :)

markstory committed:
Add API docs for EncryptedCookieMiddleware.

markstory committed:
Remove encrpytion features from individual cookies.

With an encryption middleware in place, we no longer need to have
encryption features on individual cookies. Cookies are now simple
immutable value objects and the middleware handles encryption concerns.

markstory committed:
Fix defaulted argument before required one.

markstory committed:
Fix type hints in doc comments.

markstory commented:
Working on the docs for this now.
